proposals/NNNN-spirv-push-constants.md

```diff
@@ -0,0 +1,89 @@
+---
+title: "[NNNN] - SPIR-V Push constants"
+params:
+  status: Design In Progress
+  authors:
+    - Keenuts: Nathan Gauër
+---
+
+## Introduction
+
+Push Constants are a Vulkan mechanism to pass a small amount of data to the
+shaders without creating buffers or modifying bindings.
+
+Usage is as follows:
+
+```hlsl
+struct S {
+  float3 light_color;
+};
+
+[[vk::push_constant]]
+S data;
+
+void main() {
+  [...]
+  do_something(data.light_color)
+}
+```
+## Motivation
+
+This is a core feature of HLSL+Vulkan.
+
+## Proposed Solution
+
+### HLSL - Parsing
+
+``[[vk::push_constant]]`` is a normal attribute in the `vk` namespace, with
+no custom parsing.
+This attribute is only allowed on global variables. Additional checks like
+storage class compatibility is done in sema.
+
+At this stage, we know the attribute can only be attached to a global
+variable.
+
+### HLSL - Sema
+
+During sema, compatibility checks between the global variable definition
+and the attribute are checked, and the global variable definition is created.
+
+Additional checks are:
+- The attribute is only allowed on a global variable with a struct type.
+- There can only be one ``[[vk::push_constant]]`` attribute in the shader.
+- There can be no VLA in the struct type (or a nested struct type).
+
+DXC allows any storage class to be attached to a push constant. I believe
```

Comment:
Which storage class are you referring to? Can you add examples of the syntax DXC accepts, but clang will not?